docs/proposals/cluster-autoscaling.md

```diff
@@ -0,0 +1,314 @@
+
+    // Spec defines the desired behavior of this daemon set.
+    // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+    Spec ClusterAutoscalerSpec `json:"spec,omitempty"`
+}
+```
+
+## Kubernetes 1.3 - medium KAC
+
+With medium KAC the goal is to provide a simple autoscaler that runs in Kubernetes and works in the
+similar fashion as Google Cloud Autoscaler. It will take the required metrics from Heapster.
+
+```
+                          +--Nodes---------------------------+
+                          |                                  |
+                          | +--------+ +--------+ +--------+ |
+                          | |        | |        | |        | |
+                          | | Node1  | | Node2  | | NodeX  | |
+                          | |        | |        | |        | +<--+
+                          | +--------+ +--------+ +--------+ |   |
+            +---------+   | |cAdvisor| |cAdvisor| |cAdvisor| |   |
+            |  Config |   | +--------+ +--------+ +----+---+ |   |
+            +---------+   |     |          |           |     |   |
+                 |        +-----|----------|-----------|-----+   |
+                 |              |          |           |         |
+            +----v----+         |      +---v----+      |         |
+            |   KAC   |<-+      +----> |Heapster| <----+         |
+            +---------+  |             +---+----+                |
+                 |       |                 |                     |
+                 |       +-----------------+                     |
+ Kubernetes      |                                               |
+-----------------|-----------------------------------------------|------
+ Cloud provider  |                                               |
+                 |                                               |
+                 |                                       +-------+-------+
+                 +-------------------------------------> |Node controller|
+                                                         +---------------+
+```
+
+This will solve the following problems:
+(I2) [GCE specific logic] As cluster autoscaling will have most of its components inside the cluster it will be provider independent (I2).
+
+At this point we will encourage the community to contribute code that talks to particular cloud provider node controller.
+
+## Kubernetes 1.4-1.5 - big KAC
+
+Assuming the the need for more sophisticated cluster autoscaling we will provide a full blown autoscaler that will solve (I3)[Works only for well formed pods] and will:
+
+* Take into account the number of pending pods and will check whether adding a node will allow to schedule them.
+* Choose the best node to kill once the downscale decision was made (and before will make sure that the homeless pods will schedule somewhere).
+* Reasonably handle batch-like best-effort workload
+* Use rescheduler to reduce the load.
```

Comment:
Link to https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/rescheduler.md